src/graph/mod.rs

```diff
@@ -0,0 +1,541 @@
+    }
+
+    /// Returns whether or not the graph contains a widget with the given ID.
+    pub fn contains<I: GraphIndex>(&self, idx: I) -> bool {
+        self.index_map.to_node_index(idx).is_some()
+    }
+
+
+    /// If the given Point is currently on a Widget, return an index to that widget.
+    pub fn pick_widget<I: GraphIndex>(&self, xy: Point) -> Option<I> {
+        let Graph { ref depth_order, ref graph, ref index_map, .. } = *self;
+        depth_order.iter().rev().find(|&&idx| {
+            if let Some(&Node::Widget(ref container)) = graph.node_weight(idx) {
+                if ::utils::is_over_rect(container.xy, xy, container.dim) {
+                    return true
+                }
+            }
+            false
+        }).map(|&idx| I::from_idx(idx, index_map).expect("No matching index"))
+    }
+
+
+    /// Set the parent for the given Widget Id.
+    /// This method clears all other incoming edges and ensures that the widget only has a single
+    /// parent (incoming edge). This means we can be sure of retaining a tree structure.
+    pub fn set_parent_for_widget<I, P>(&mut self, idx: I, maybe_parent_idx: Option<P>) where
+        I: GraphIndex,
+        P: GraphIndex,
+    {
+        let Graph { ref mut graph, ref mut index_map, root, .. } = *self;
+        let node_idx = index_map.to_node_index(idx).expect("No NodeIndex for given GraphIndex");
+        // If no parent id was given, we will set the root as the parent.
+        let parent_node_idx = match maybe_parent_idx {
+            Some(parent_idx) => match index_map.to_node_index(parent_idx) {
+                Some(parent_node_idx) => parent_node_idx,
+                // Add a temporary node to the graph at the given parent index so that we can add
+                // the edge even in the parent widget's absense. The temporary node should be
+                // replaced by the proper widget when it is updated later in the cycle.
+                None => {
+                    // We *know* that this must be a WidgetId as `.to_node_indx` returned None.
+                    let parent_widget_id = parent_idx.expect_widget_id();
+                    // Add a placeholder node to act as a parent until the actual parent is placed.
+                    let parent_node_idx = graph.add_node(Node::Placeholder);
+                    index_map.insert(parent_widget_id, parent_node_idx);
+                    parent_node_idx
+                },
+            },
+            None => root,
+        };
+
+        // Walk our incoming edges and remove any that aren't connected to our requested parent.
+        let mut incoming_edges = graph.walk_edges_directed(node_idx, pg::Incoming);
+        let mut already_connected = false;
+        while let Some((in_edge_idx, in_node_idx)) = incoming_edges.next_neighbor(graph) {
+            if in_node_idx == parent_node_idx {
+                already_connected = true;
+            } else {
+                graph.remove_edge(in_edge_idx);
```

Comment:
@bluss to the rescue once again!

As it turned out, there can only ever be one edge at a time (as that method is the only publicly available way of adding edges) so it can't cause any errors as it stands, however it's certainly not clear and could easily cause issues in the future when refactoring etc!

I might just change this from `while let` to `if let` for now and add some decent docs.

Thanks for being lovely and looking out for your reverse deps!